test/common/test_context.go

```diff
@@ -80,76 +127,71 @@ func TestEnv(additionalFlagFuncs ...func(set *pflag.FlagSet)) env.Environment {
+	for name, server := range additionalFakeServers {
+		tcb.Servers[name] = server
+	}
 
-	flag.VisitAll(func(flag *flag.Flag) {
-		if flag.Value.String() != "" {
-			// if any of the go test flags have been set, propagate the value in sm env with highest prio
-			// when env exposes the pflagset it would be better to instead override the pflag value instead
-			smEnv.Set(flag.Name, flag.Value.String())
-		}
-	})
+	return tcb
+}
 
-	ctx, cancel := context.WithCancel(context.Background())
-	s := struct {
-		Log *log.Settings
-	}{
-		Log: &log.Settings{
-			Output: ginkgo.GinkgoWriter,
-		},
-	}
-	err := smEnv.Unmarshal(&s)
-	if err != nil {
-		panic(err)
-	}
-	ctx = log.Configure(ctx, s.Log)
-	smanagerBuilder := sm.New(ctx, cancel, smEnv)
-	if params.RegisterExtensions != nil {
-		params.RegisterExtensions(smanagerBuilder)
-	}
-	serviceManager := smanagerBuilder.Build()
-	return httptest.NewServer(serviceManager.Server.Router)
+func (tcb *TestContextBuilder) WithDefaultTokenClaims(defaultTokenClaims map[string]interface{}) *TestContextBuilder {
+	tcb.defaultTokenClaims = defaultTokenClaims
+
+	return tcb
 }
 
-func NewTestContext(params *ContextParams) *TestContext {
-	if params == nil {
-		params = &ContextParams{}
-	}
+func (tcb *TestContextBuilder) WithEnvPreExtensions(fs ...func(set *pflag.FlagSet)) *TestContextBuilder {
+	tcb.envPreHooks = append(tcb.envPreHooks, fs...)
+
+	return tcb
+}
+
+func (tcb *TestContextBuilder) WithEnvPostExtensions(fs ...func(e env.Environment, servers map[string]FakeServer)) *TestContextBuilder {
+	tcb.envPostHooks = append(tcb.envPostHooks, fs...)
+
+	return tcb
+}
 
-	oauthServer := NewOAuthServer()
-	oauthServer.Start()
+func (tcb *TestContextBuilder) WithSMExtensions(fs ...func(ctx context.Context, smb *sm.ServiceManagerBuilder, e env.Environment) error) *TestContextBuilder {
+	tcb.smExtensions = append(tcb.smExtensions, fs...)
+
+	return tcb
+}
+
+func (tcb *TestContextBuilder) Build() *TestContext {
+	environment := tcb.Environment(tcb.envPreHooks...)
+
+	for _, envPostHook := range tcb.envPostHooks {
+		envPostHook(environment, tcb.Servers)
+	}
 
-	smServer := NewSMServer(params, oauthServer.URL)
-	SM := httpexpect.New(GinkgoT(), smServer.URL)
+	smServer := newSMServer(environment, tcb.smExtensions)
+	tcb.Servers[SMServer] = smServer
 
-	accessToken := oauthServer.CreateToken(params.DefaultTokenClaims)
+	SM := httpexpect.New(GinkgoT(), smServer.URL())
+	oauthServer := tcb.Servers[OauthServer].(*OAuthServer)
+	accessToken := oauthServer.CreateToken(tcb.defaultTokenClaims)
 	SMWithOAuth := SM.Builder(func(req *httpexpect.Request) {
 		req.WithHeader("Authorization", "Bearer "+accessToken)
 	})
 	RemoveAllBrokers(SMWithOAuth)
 	RemoveAllPlatforms(SMWithOAuth)
 
-	platformJSON := MakePlatform("ctx-platform-test", "ctx-platform-test", "platform-type", "test-platform")
+	platformJSON := MakePlatform("tcb-platform-test", "tcb-platform-test", "platform-type", "test-platform")
```

Comment:
can the platform that is created be extracted in another step? Or have a `builder.SkipResourceCreation`

Reply:
yes we can do that